### PR TITLE
pyside 5.12.2

### PR DESCRIPTION
**Pull request by albertosottile**
Bumps to the latest version of `pyside`.
Since this version, it is possible again to use the standard `Language::Python.setup_install_args` wrapper to install `pyside`. So, this commit restores this installation approach, as previously
discussed here: https://github.com/Homebrew/homebrew-core/pull/35910#issuecomment-461097403.
In addition, the dependency from `llvm` is now limited to build time, avoiding the need to download its huge bottle to use `pyside`. Tests were adjusted accordingly.

- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/master/CONTRIBUTING.md)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?

-----

### CODE REVIEW

File: Formula/pyside.rb [hunk 1]
```diff
@@ -11,38 +11,39 @@ class Pyside < Formula
   end
 
   depends_on "cmake" => [:build, :test]
-  depends_on "llvm"
+  depends_on "llvm" => :build
```
Comment (fxcoudert) [inline Formula/pyside.rb]: Why is it using `llvm` for, if it is only for build?

File: Formula/pyside.rb [hunk 2]
```diff
@@ -11,38 +11,39 @@ class Pyside < Formula
...
   depends_on "python"
   depends_on "python@2"
   depends_on "qt"
 
   def install
+    ENV.remove "HOMEBREW_LIBRARY_PATHS", Formula["llvm"].opt_lib
```
Comment (fxcoudert) [inline Formula/pyside.rb]: Why is this needed? It looks like you are trying to use `llvm` but not link against it, which is wrong. We want to avoid mismatched compiler/runtime.

File: Formula/pyside.rb [hunk 3]
```diff
@@ -11,38 +11,39 @@ class Pyside < Formula
...
-           "--install-lib", lib/"python#{xy}/site-packages"
+    system "python3", *Language::Python.setup_install_args(prefix),
+           "--install-lib", lib/"python#{xy}/site-packages", *args
 
     lib.install_symlink Dir.glob(lib/"python#{xy}/site-packages/PySide2/*.dylib")
     lib.install_symlink Dir.glob(lib/"python#{xy}/site-packages/shiboken2/*.dylib")
 
-    system "python2", "setup.py", *args,
-           "--install-lib", lib/"python2.7/site-packages"
+    rm_rf bin/"shiboken2"
```
Comment (fxcoudert) [inline Formula/pyside.rb]: If it is installed by `setup.py` by default, we probably don't want to remove it.

File: Formula/pyside.rb [hunk 4]
```diff
@@ -61,18 +62,5 @@ def install
         Xml
       ].each { |mod| system python, "-c", "import PySide2.Qt#{mod}" }
     end
-    ["python", "python@2"].each do |python|
```
Comment (fxcoudert) [inline Formula/pyside.rb]: Why is that part of the test removed?

**Review by fxcoudert — CHANGES_REQUESTED** [review]

File: Formula/pyside.rb (hunk 1 above)
Comment (albertosottile) [reply Formula/pyside.rb]: `pyside` uses libclang to parse `qt` headers and generate the python bindings at compile time. libclang is not required by `pyside` at runtime. Initially, we had to include `llvm` because of opportunistic linkage of `libc++.1.dylib` from it, instead that from system. This creates, of course, a potential liability because this approach differs from upstream, which links to the library provided by the system. So, with @ilovezfs, we had an agreement to remove this runtime dependency as soon as possible (see the [5.11.0 PR](https://github.com/Homebrew/homebrew-core/pull/29699) or the [proposed then aborted PR](url
https://github.com/Homebrew/homebrew-core/pull/30229) for `libclang` for further info).

File: Formula/pyside.rb (hunk 2 above)
Comment (albertosottile) [reply Formula/pyside.rb]: This is needed to avoid linking `pyside` to the `libc++.1.dylib` provided by `llvm`.

File: Formula/pyside.rb (hunk 3 above)
Comment (albertosottile) [reply Formula/pyside.rb]: The shiboken binary is not a part of `pyside` and is not required for using the binding. Indeed, the official distribution of `pyside` does not include it. The reason why it was not removed before was to allow a test of the correct linkage of `llvm`, which was opportunistically linked in a way that it was not possible to avoid, until now.

File: Formula/pyside.rb (hunk 4 above)
Comment (albertosottile) [reply Formula/pyside.rb]: As I said before, importing the python bindings worked without `llvm` but, due to the opportunistic linkage, `llvm` was still required and a separate test was needed to check if this linkage was correct. So, I had to create a shiboken2-based test (see [here](https://github.com/Homebrew/homebrew-core/pull/29699/commits/0d607c220ba4cd3cd70e2a791d7a63aa87301734)).  Now the linkage does not require `llvm` anymore, so there is no need to embed shiboken in the `pyside` formula, hence this test is not required.

**Comment by albertosottile** [comment]
I'll try to wrap things up here: these adjustments to the formula mimic those performed upstream, essentially eliminating the need to ship libclang (or `llvm`) libraries with pyside. Upstream, they separated the `pyside` package in three sub packages (pyside2, shiboken2, and shiboken-generator). The first two do not depend on libclang, while the last does. Only the last contains the shiboken2 binary and `libclang.dylib`. But, the upstream installation of pyside includes only the first two packages. 

So, with these adjustments we can achieve the same result and provide a formula for `pyside` that provides exactly the same features of the upstream package, and additionally removes the runtime dependency from `llvm`.

File: Formula/pyside.rb (hunk 3 above)
Comment (fxcoudert) [reply Formula/pyside.rb]: I do not understand:

> The shiboken binary is not a part of pyside

and 

> the official distribution of pyside does not include it

We are using the official source distribution, are we not? It appears to be included, and installed by default. If this is not the desired behaviour, it should be reported upstream as a bug.

Homebrew's policy is that we want to ship things as close to possible as the default building configuration. Every deviation from that is possible cause for breakage and extra maintenance burden down the road.

Comment (albertosottile) [reply Formula/pyside.rb]: The official installation method is based on pre-built wheels loaded from pypi, via running `pip install PySide2`. You can verify that installations done via this method do not include a shiboken2 nor a `libclang.dylib` binary. 

Installing from the source archive creates and installs a shiboken2 binary, instead, essentially because that binary is required to build pyside, but not to run it. That shiboken2 binary is linked to libclang (from `llvm`) and so it was removed in this formula.

Comment (albertosottile) [reply Formula/pyside.rb]: If you are particularly concerned about the manual removals, I just realized and tested that it is possible to achieve the same result by running 
```
    system "python3", "setup.py", *args,
           "--install-lib", lib/"python#{xy}/site-packages",
           "--build-type=shiboken2"
           
    system "python3", "setup.py", *args,
           "--install-lib", lib/"python#{xy}/site-packages",
           "--build-type=pyside2"
```
for each python version, instead of the previous single `setup.py` invocation.

Does this sound any better?

**Comment by albertosottile** [comment]
@fxcoudert Any news on this? Do you prefer if I implement the changes suggested [here](https://github.com/Homebrew/homebrew-core/pull/38180#discussion_r268367453) (in short: removing the manual `rm` calls and slightly amend the build commands to achieve the same result)?

**Comment by albertosottile** [comment]
@fxcoudert Ping. Is there anything I can do to move this forward?

**Comment by albertosottile** [comment]
Ping. Soon 5.12.3 will be released... I have been waiting for an answer since more than a month now.

**Comment by albertosottile** [comment]
Ping. Soon 5.12.3 will be released... I have been waiting for an answer since more than a month now. @fxcoudert @SMillerDev

**Comment by albertosottile** [comment]
Pyside 5.12.3 has been released. I bumped this formula to the new version locally, and tests are working fine. Should I update this PR or close it for good and open a new one for the new version? Honestly, at this point any feedback would be much appreciated...

**Comment by albertosottile** [comment]
@fxcoudert @SMillerDev @MikeMcQuaid Ping. Is there anything I can do? Should I close this PR and open a new one for 5.12.3?

File: Formula/pyside.rb (hunk 3 above)
Comment (MikeMcQuaid) [reply Formula/pyside.rb]: > Does this sound any better?

Yes, please!

**Comment by MikeMcQuaid** [comment]
@albertosottile I have one pending comment based on your suggestion. After that's addressed and this PR is updated to the new version I'm happy to pull it 👍

**Comment by albertosottile** [comment]
> @albertosottile I have one pending comment based on your suggestion. After that's addressed and this PR is updated to the new version I'm happy to pull it 👍

Thanks for addressing this. I just updated this PR as you suggested. The branch name is still on the old version, though.

**Review by MikeMcQuaid — APPROVED** [review]

**Comment by MikeMcQuaid** [comment]
@fxcoudert Are you 🆗 with this as-is?

**Review by fxcoudert — APPROVED** [review]

**Comment by MikeMcQuaid** [comment]
Thanks again so much for your contribution! Without people like you submitting PRs we couldn't run this project. You rock, @albertosottile.
